jsonrpc/ipc.go

```diff
@@ -0,0 +1,198 @@
+package jsonrpc
+
+import (
+	"context"
+	"errors"
+	"io"
+	"net"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/NethermindEth/juno/utils"
+	"github.com/ethereum/go-ethereum/p2p/netutil"
+	"github.com/sourcegraph/conc"
+)
+
+type Ipc struct {
+	endpoint string
+	rpc      *Server
+	events   NewRequestListener
+	log      utils.SimpleLogger
+
+	connWg sync.WaitGroup // connWg is a WaitGroup for tracking active connections.
+
+	connParams IpcConnParams
+	listener   net.Listener
+
+	socketCallback func() // socketCallback is executed when socket has been created
+
+	// everything below is protected
+	mu      sync.Mutex
+	conns   map[net.Conn]struct{} // conns is a map that holds active connections.
+	closing bool                  // closing indicates whether the server is in the process of shutting down.
+}
+
+// NewIpc creates a new IPC handler instance with the provided RPC server, endpoint and logger.
+func NewIpc(rpc *Server, endpoint string, log utils.SimpleLogger) *Ipc {
+	return &Ipc{
+		endpoint:       endpoint,
+		rpc:            rpc,
+		log:            log,
+		connParams:     DefaultIpcConnParams(),
+		conns:          make(map[net.Conn]struct{}),
+		events:         &SelectiveListener{},
+		socketCallback: func() {},
+	}
+}
+
+// WithSocketCallback sets the callback for socket creation.
+func (i *Ipc) WithSocketCallback(f func()) *Ipc {
+	i.socketCallback = f
+	return i
+}
+
+// WithConnParams sanity checks and applies the provided params.
+func (i *Ipc) WithConnParams(p *IpcConnParams) *Ipc {
+	i.connParams = *p
+	return i
+}
+
+// WithListener registers a NewRequestListener
+func (i *Ipc) WithListener(listener NewRequestListener) *Ipc {
+	i.events = listener
+	return i
+}
+
+// Run launches the IPC handler and handles any potential errors.
+// It is the caller's responsibility to cancel the provided context when they wish to
+// gracefully shut down the IPC handler.
+func (i *Ipc) Run(ctx context.Context) error {
+	listener, err := createListener(i.endpoint)
+	if err != nil {
+		return err
+	}
+	i.listener = listener
+	i.socketCallback()
+	var (
+		ch = make(chan struct{})
+		wg conc.WaitGroup
+	)
+	defer wg.Wait()
+	wg.Go(func() {
+		defer close(ch)
+		for {
+			conn, err := i.listener.Accept()
+			if netutil.IsTemporaryError(err) {
+				i.log.Warnw("Failed to accept connection", "err", err)
+				continue
+			} else if err != nil {
+				i.log.Warnw("Accept connection", "err", err)
+				return
+			}
+			i.connWg.Add(1)
+			go i.serveConn(ctx, newIpcConn(conn, i.connParams))
+		}
+	})
+
+	select {
+	case <-ctx.Done():
+	case <-ch:
```

Comment:
Rather than managing a separate `ch` channel, can we use a child context of `ctx`?

Reply:
sure